# Multiomics_integrative_analysis

A multi-layered multiomic approach to integrate chromatin analysis (Ma et al) at all accessible regions across wildtype and mutant HSPC was applied. Firstly, within 2-kb bins at ATAC-seq consensus peaks (Â±1 kb from peak summit, n=75,457), normalised sequencing read counts for each chromatin modifications (H3K4me1, H3K4me3 and H3K27ac) and chromatin accessibility of each HSPC population (WT, Npm1c, Flt3-ITD and DM) were extracted and computed as input data matrix. This matrix was composed of 16 rows (with 4x chromatin marks and 4x cell populations) and 75,457 columns (with bins of accessible regions). Then processing the data in a similar way as for single cell RNA-sequencing with Seurat package, all 75,457 accessible regions (as columns) were treated as a separate data point across all 16 conditions (as rows). This allowed dimensionality reduction to classify and visualise clusters of regions whose pattern of alterations were similar. Data were plotted on a tSNE plot to visualize all separated clusters (n=10) and a heatmap was shown to illustrate the specific patterns of chromatin changes in each cell type (no row-wise ranking). These clusters were further loaded as regions of interest into deepTools to probe average profiles of chromatin modifications and accessibility across WT and mutant conditions, further supporting the separation of tSNE clusters and their unique changes of chromatin states. 

Multiomics_Seurat_analysis.R

Ma, Q. et al. Super-Enhancer Redistribution as a Mechanism of Broad Gene Dysregulation in Repeatedly Drug-Treated Cancer Cells. Cell Rep. 31, 107532 (2020).
